---
title: "Untitled"
author: "Mthande_Fezile"
date: "8/14/2020"
output: pdf_document
---

Importing our data
```{r}
sample= read.table("D:/Matrix/counts1.Rmatrix.txt", header= TRUE, sep="\t", row.names = 1)
```
renaming the headers of the file
```{r}
colnames(sample) <- gsub("Bam", "", colnames(sample))
colnames(sample) <- gsub(".sorted.bam", "", colnames(sample))
colnames(sample) <-gsub("\\.", "" , colnames(sample))
head(sample)
```
importing the metadata

```{r}
meta= read.table("D:/Matrix/practice.dataset.metadata.tsv", header=TRUE, sep="\t", row.names = 1)
head(meta)
meta$Condition <- factor(meta$Condition)
```

confirming if all the colnames correspond
```{r}
all(rownames(meta) %in% colnames(sample))
all(rownames(meta) == colnames(sample))
```

```{r,include=FALSE}
library("DESeq2")
```
creating the dds
```{r}
dds= DESeqDataSetFromMatrix(countData = sample, colData = meta, design = ~ Condition)

dds
```

combining
```{r}
info <- as.data.frame( colData( dds))
head(info)
```

Filtering rows with low reads(>=10)
```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
dds
```

specifying the levels
```{r}
dds$Condition <- factor(dds$Condition, levels = c("disease","normal"))
```

running deseq
```{r}
dds <- DESeq(dds)
res <- results(dds)
res
```

specifying comparison of interest using contrast

```{r}
res <- results(dds, contrast=c("Condition","normal","disease"))
res
```

shrinkage and ranking of genes
```{r}
library("apeglm")
resLFC <- lfcShrink(dds, coef="Condition_normal_vs_disease", type="apeglm")
resLFC
```

ordering according to p-value
```{r}
resOrdered <- res[order(res$pvalue),]
```

genes with the smalles pvalues
```{r}
idx <- which.min(res$pvalue)
counts(dds)[idx, ]
```


summarizing
```{r}
summary(res)
```

p-values less than 0.1/10% acceptable(FDR)
```{r}
sum(res$padj < 0.1, na.rm=TRUE)
```

setting alpha
```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)

sum(res05$padj < 0.05, na.rm=TRUE)
```

INDEPENDENT HYPOTHESIS TESTING
```{r, results='hide'}
library(IHW)
```
```{r}
resIHW <- results(dds, filterFun=ihw)
summary(resIHW)
sum(resIHW$padj < 0.1, na.rm=TRUE)
metadata(resIHW)$ihwResult
```


genes with high upregulation
```{r}
resSig <- subset(res, res$padj < 0.1 )
head( resSig[ order( -resSig$log2FoldChange ), ],10)
```

genes with down regulation
```{r}
head( resSig[ order( resSig$log2FoldChange ), ],10)
```

PLOTS


#EXPLORATION AND EXPORTING DATA
1.MA-PLOTS

```{r}
plotMA(res, ylim=c(-2,2))
```

SHRUNKEN FOLD CHAGES
```{r}
plotMA(resLFC, ylim=c(-2,2))
```

#IDENTIFYING THE NUMBER OF GENES
#idx <- identify(res$baseMean, res$log2FoldChange)
#rownames(res)[idx]


SPECIFYING THE COEFFICIENT 2
```{r}
resultsNames(dds)
resNorm <- lfcShrink(dds, coef = 2, type="normal")
resNorm <- lfcShrink(dds, coef = 2, type="ashr")
```

```{r, results="hide"}
library("ashr")
```

the plots
```{r}
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resLFC, xlim=xlim, ylim=ylim, main="ashr")
```


PLOT COUNTS
```{r}
plotCounts(dds, gene=which.min(res$padj), intgroup="Condition")


#using ggplot

d <- plotCounts(dds, gene=which.min(res$padj), intgroup="Condition", 
                returnData=TRUE)
library("ggplot2")
ggplot(d, aes(x=Condition, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))


#results columns


mcols(res)$description


```
2.VOLCANO PLOT

reset par
```{r}
par(mfrow=c(1,1))

#Make a basic volcano plot

with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-3,3)))


#Add colored points: blue if padj<0.01, red if log2FC>1 and padj<0.05)

with(subset(res, padj<.01 ), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, padj<.01 & abs(log2FoldChange)>2), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
```

PCA
```{r}
vsdata <- vst(dds, blind=FALSE)
plotPCA(vsdata, intgroup="Condition") #using the DESEQ2 plotPCA fxn we can
```

ANNOTATION

ADDING GENE NAMES
```{r, results='hide'}
library( "org.Hs.eg.db" )
```

get the list of key types
```{r}
columns(org.Hs.eg.db)
```
easy function for conversion
```{r}
convertIDs <- function( ids, fromKey, toKey, db, ifMultiple=c( "putNA", "useFirst" ) ) {
  stopifnot( inherits( db, "AnnotationDb" ) )
  ifMultiple <- match.arg( ifMultiple )
  suppressWarnings( selRes <- AnnotationDbi::select( 
    db, keys=ids, keytype=fromKey, columns=c(fromKey,toKey) ) )
  if( ifMultiple == "putNA" ) {
    duplicatedIds <- selRes[ duplicated( selRes[,1] ), 1 ]   
    selRes <- selRes[ ! selRes[,1] %in% duplicatedIds, ] }
  return( selRes[ match( ids, selRes[,1] ), 2 ] )
}
```

converting the ideas into the names and symbols founf in the Annotation databse
```{r}
res$hgnc_symbol <- convertIDs( row.names(res), "ENSEMBL", "SYMBOL", org.Hs.eg.db )
res$entrezid <- convertIDs( row.names(res), "ENSEMBL", "ENTREZID", org.Hs.eg.db )
tail(res)

resSig$hgnc_symbol <- convertIDs( row.names(resSig), "ENSEMBL", "SYMBOL", org.Hs.eg.db )
tail(resSig)
```

GENES THAT ARE UP REGULATED WITH THIER NAMES AND SYMBOL
```{r}
up_reg = head( resSig[ order( -resSig$log2FoldChange ), ],10)
up_reg$hgnc_symbol <- convertIDs( row.names(up_reg), "ENSEMBL", "SYMBOL", org.Hs.eg.db )
up_reg$entrezid <- convertIDs( row.names(up_reg), "ENSEMBL", "ENTREZID", org.Hs.eg.db )
head(up_reg)
```

Low regulated genes
```{r}
down_reg = head( resSig[ order( -resSig$log2FoldChange ), ],10)
down_reg$hgnc_symbol <- convertIDs( row.names(down_reg), "ENSEMBL", "SYMBOL", org.Hs.eg.db )
down_reg$entrezid <- convertIDs( row.names(down_reg), "ENSEMBL", "ENTREZID", org.Hs.eg.db )
head(down_reg)
```

Dispersion plot
```{r}
plotDispEsts( dds, ylim = c(1e-6, 1e1) )
```

HISTOGRAM OF P -VALUES
```{r}
hist( res$pvalue, breaks=20, col="red" )
```